File: leetcode/solutions/0677.MapSumPairs/MapSum.ts
export interface TrieNode {
  val: number;
  children: Map<string, TrieNode>;
}

export default class MapSum {
  private trie: TrieNode;

  constructor() {
    this.trie = this.createTrieNode();
  }

  insert(key: string, val: number) {
    let node = this.trie;

    for (const char of key) {
      let nextNode = node.children.get(char);

      if (!nextNode) {
        nextNode = this.createTrieNode();
        node.children.set(char, nextNode);
      }

      node = nextNode;
    }

    node.val = val;
  }

  sum(prefix: string): number {
    let node = this.trie;

    for (const char of prefix) {
      const nextNode = node.children.get(char);

      if (!nextNode) {
        return 0;
      }

      node = nextNode;
    }

    function sumUp(node: TrieNode | null): number {
      if (!node) {
        return 0;
      }

      let sum = node.val;

      for (const child of node.children.values()) {
        sum += sumUp(child);
      }

      return sum;
    }

    return sumUp(node);
  }

  createTrieNode(): TrieNode {
    return { val: 0, children: new Map<string, TrieNode>() };
  }
}

File: leetcode/solutions/0677.MapSumPairs/MapSum.go
package solution

type MapSum struct {
	num      int
	children [26]*MapSum
}

// Constructor initializes a MapSum data.
func Constructor() MapSum {
	return MapSum{}
}

// Insert adds a word into a trie.
func (m *MapSum) Insert(key string, val int) {
	node := m

	for _, r := range key {
		key := r - 'a'

		if node.children[key] == nil {
			node.children[key] = &MapSum{}
		}

		node = node.children[key]
	}

	node.num = val
}

// Sum adds up all the values prefixed with the given string.
func (m *MapSum) Sum(prefix string) int {
	node := m

	for _, r := range prefix {
		key := r - 'a'

		if node.children[key] == nil {
			return 0
		}

		node = node.children[key]
	}

	return sumNodes(node)
}

// sumNodes adds up all the values in the given trie node.
func sumNodes(node *MapSum) int {
	if node == nil {
		return 0
	}

	sum := node.num

	for _, child := range node.children {
		sum += sumNodes(child)
	}

	return sum
}
